File: memory_sharer/race_condition.rb
$acumulador = 0
repeticoes = 100_000

def incrementa
  sleep(0.0001)
  $acumulador += 1
end

def decrementa
  sleep(0.0001)
  $acumulador -= 1
end

def incrementador(repeticoes)
  repeticoes.times do
    incrementa
  end
end

def decrementador(repeticoes)
  repeticoes.times do
    decrementa
  end
end

thread1 = Thread.new { incrementador(repeticoes) }
thread2 = Thread.new { decrementador(repeticoes) }

thread1.join
thread2.join

puts "Resultado final: #{$acumulador}"

File: memory_sharer/race_condition.py
from threading import Thread

acumulador = 0
repeticoes = 1_000_000

def incrementa():
    global acumulador
    acumulador += 1

def decrementa():
    global acumulador
    acumulador -= 1

def incrementador(repeticoes: int):
    for _ in range(repeticoes):
        incrementa()

def decrementador(repeticoes: int):
    for _ in range(repeticoes):
        decrementa()

if __name__ == "__main__":
    thread1 = Thread(target=incrementador, args=(repeticoes,))
    thread2 = Thread(target=decrementador, args=(repeticoes,))

    thread1.start()
    thread2.start()

    thread1.join()
    thread2.join()

    print(f'Resultado Final: {acumulador}')
